File: routed-to-an-out-of-stock-location/python/reroute_out_of_stock_fulfillment.py
"""Move Shopify fulfillment orders off a location that cannot stock them.

An order can land on a location that shows OPEN or IN_PROGRESS but has no
usable inventory for one or more line items there, often because a location
rule or the customer's address routed it there before a stock count caught
up. Shopify will not fulfill from a location with nothing to pick, so the
order stalls. This lists fulfillment orders assigned to a "problem" location,
asks Shopify which other locations could take the line items with
locationsForMove, picks the best candidate in pure code, and calls
fulfillmentOrderMove to reassign it. Read only apart from the move.
Run on a schedule. Safe to run again and again.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reroute_out_of_stock_fulfillment")

SHOP = os.environ["SHOPIFY_SHOP"]
TOKEN = os.environ["SHOPIFY_ACCESS_TOKEN"]
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
PROBLEM_LOCATION_ID = os.environ["OUT_OF_STOCK_LOCATION_ID"]
MOVABLE_STATUSES = {"OPEN", "IN_PROGRESS"}
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

FULFILLMENT_ORDERS_QUERY = """
query($cursor: String, $locationId: ID!) {
  location(id: $locationId) {
    fulfillmentOrders(first: 25, after: $cursor) {
      pageInfo { hasNextPage endCursor }
      nodes {
        id
        status
        lineItems(first: 50) { nodes { id remainingQuantity } }
        locationsForMove(first: 10) {
          nodes {
            location { id name }
            message
            availableLineItems(first: 50) { nodes { id remainingQuantity } }
          }
        }
      }
    }
  }
}"""

MOVE_MUTATION = """
mutation($id: ID!, $newLocationId: ID!) {
  fulfillmentOrderMove(id: $id, newLocationId: $newLocationId) {
    movedFulfillmentOrder { id status }
    userErrors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def total_remaining(line_items):
    return sum(item.get("remainingQuantity", 0) for item in line_items or [])


def pick_reroute_location(fulfillment_order):
    """Pure decision: which location (if any) should this fulfillment order move to.

    Returns a candidate location id, or None if the order should be left alone.
    The order is only a candidate when it is still movable (OPEN or IN_PROGRESS)
    and it has unfulfilled line items. Among the locations Shopify reports with
    locationsForMove, we keep only the ones that can cover every remaining line
    item (no partial moves that would split the order in two), and we pick the
    one that covers the most remaining quantity so the busiest order clears
    first when several locations qualify.
    """
    if fulfillment_order.get("status") not in MOVABLE_STATUSES:
        return None

    needed = total_remaining(fulfillment_order.get("lineItems", {}).get("nodes"))
    if needed <= 0:
        return None

    best_id = None
    best_covered = -1
    for candidate in fulfillment_order.get("locationsForMove", {}).get("nodes", []):
        covered = total_remaining(candidate.get("availableLineItems", {}).get("nodes"))
        if covered < needed:
            continue
        if covered > best_covered:
            best_covered = covered
            best_id = candidate["location"]["id"]

    return best_id


def stuck_fulfillment_orders():
    cursor = None
    while True:
        data = gql(FULFILLMENT_ORDERS_QUERY, {"cursor": cursor, "locationId": PROBLEM_LOCATION_ID})
        location = data.get("location")
        if location is None:
            return
        page = location["fulfillmentOrders"]
        for node in page["nodes"]:
            yield node
        if not page["pageInfo"]["hasNextPage"]:
            return
        cursor = page["pageInfo"]["endCursor"]


def move_fulfillment_order(fulfillment_order_id, new_location_id):
    result = gql(MOVE_MUTATION, {"id": fulfillment_order_id, "newLocationId": new_location_id})["fulfillmentOrderMove"]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])
    return result["movedFulfillmentOrder"]["status"]


def run():
    moved = 0
    for fulfillment_order in stuck_fulfillment_orders():
        target = pick_reroute_location(fulfillment_order)
        if target is None:
            continue
        log.info(
            "Fulfillment order %s can move to %s. %s",
            fulfillment_order["id"], target, "would move" if DRY_RUN else "moving",
        )
        if not DRY_RUN:
            move_fulfillment_order(fulfillment_order["id"], target)
        moved += 1
    log.info("Done. %d fulfillment order(s) %s.", moved, "to move" if DRY_RUN else "moved")


if __name__ == "__main__":
    run()

File: routed-to-an-out-of-stock-location/node/reroute-out-of-stock-fulfillment.js
/**
 * Move Shopify fulfillment orders off a location that cannot stock them.
 *
 * An order can land on a location that shows OPEN or IN_PROGRESS but has no
 * usable inventory for one or more line items there, often because a location
 * rule or the customer's address routed it there before a stock count caught
 * up. Shopify will not fulfill from a location with nothing to pick, so the
 * order stalls. This lists fulfillment orders assigned to a "problem"
 * location, asks Shopify which other locations could take the line items
 * with locationsForMove, picks the best candidate in pure code, and calls
 * fulfillmentOrderMove to reassign it. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/shopify/routed-to-an-out-of-stock-location/
 */
import { pathToFileURL } from "node:url";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const PROBLEM_LOCATION_ID = process.env.OUT_OF_STOCK_LOCATION_ID || "gid://shopify/Location/1";
const MOVABLE_STATUSES = new Set(["OPEN", "IN_PROGRESS"]);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function totalRemaining(lineItems) {
  return (lineItems || []).reduce((sum, item) => sum + (item.remainingQuantity || 0), 0);
}

/**
 * Pure decision: which location (if any) should this fulfillment order move to.
 *
 * Returns a candidate location id, or null if the order should be left alone.
 * The order is only a candidate when it is still movable (OPEN or IN_PROGRESS)
 * and it has unfulfilled line items. Among the locations Shopify reports with
 * locationsForMove, we keep only the ones that can cover every remaining line
 * item (no partial moves that would split the order in two), and we pick the
 * one that covers the most remaining quantity so the busiest order clears
 * first when several locations qualify.
 */
export function pickRerouteLocation(fulfillmentOrder) {
  if (!MOVABLE_STATUSES.has(fulfillmentOrder.status)) return null;

  const needed = totalRemaining(fulfillmentOrder.lineItems?.nodes);
  if (needed <= 0) return null;

  let bestId = null;
  let bestCovered = -1;
  for (const candidate of fulfillmentOrder.locationsForMove?.nodes || []) {
    const covered = totalRemaining(candidate.availableLineItems?.nodes);
    if (covered < needed) continue;
    if (covered > bestCovered) {
      bestCovered = covered;
      bestId = candidate.location.id;
    }
  }
  return bestId;
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const FULFILLMENT_ORDERS_QUERY = `
query($cursor: String, $locationId: ID!) {
  location(id: $locationId) {
    fulfillmentOrders(first: 25, after: $cursor) {
      pageInfo { hasNextPage endCursor }
      nodes {
        id
        status
        lineItems(first: 50) { nodes { id remainingQuantity } }
        locationsForMove(first: 10) {
          nodes {
            location { id name }
            message
            availableLineItems(first: 50) { nodes { id remainingQuantity } }
          }
        }
      }
    }
  }
}`;

const MOVE_MUTATION = `
mutation($id: ID!, $newLocationId: ID!) {
  fulfillmentOrderMove(id: $id, newLocationId: $newLocationId) {
    movedFulfillmentOrder { id status }
    userErrors { field message }
  }
}`;

async function* stuckFulfillmentOrders() {
  let cursor = null;
  while (true) {
    const data = await gql(FULFILLMENT_ORDERS_QUERY, { cursor, locationId: PROBLEM_LOCATION_ID });
    const location = data.location;
    if (!location) return;
    const page = location.fulfillmentOrders;
    for (const node of page.nodes) yield node;
    if (!page.pageInfo.hasNextPage) return;
    cursor = page.pageInfo.endCursor;
  }
}

async function moveFulfillmentOrder(fulfillmentOrderId, newLocationId) {
  const result = (await gql(MOVE_MUTATION, { id: fulfillmentOrderId, newLocationId })).fulfillmentOrderMove;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
  return result.movedFulfillmentOrder.status;
}

export async function run() {
  let moved = 0;
  for await (const fulfillmentOrder of stuckFulfillmentOrders()) {
    const target = pickRerouteLocation(fulfillmentOrder);
    if (!target) continue;
    console.log(`Fulfillment order ${fulfillmentOrder.id} can move to ${target}. ${DRY_RUN ? "would move" : "moving"}`);
    if (!DRY_RUN) await moveFulfillmentOrder(fulfillmentOrder.id, target);
    moved++;
  }
  console.log(`Done. ${moved} fulfillment order(s) ${DRY_RUN ? "to move" : "moved"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
